Add bindgen build zephyr-sys

d3zd3z opened the pull request:
With a lot of help and inspiration from @mjaun, generate a zephyr-sys crate using bindgen.

This currently exposes Zpehyr functions starting with `k_` and with `gpio_`, although the code currently only uses `k_str_out` for the `printk/printkln` implementation.

The bindgen tool generates a wrapper C file that has wrappers for functions declared as static inline in the Zephyr headers. Although there will be a bit of an efficiency hit, this gives us a stable way of accessing the Zephyr interfaces. If there are any that would be better handled directly in Rust, they can be done by hand, and bindgen instructed to skip them.

d3zd3z committed:
rust: Create zephyr-sys for low-level bindings

With a lot of work done by Mario Jaun <mario.jaun@zuehlke.com>.

Use the rust-bindgen tool to extract bindings from the Zephyr header files.
The tool is run with the exact configuration of the current build, and
therefore the bindings will match the current target.  The ifdefs and such
are not translated into the Rust code, and this must be generated live for
each build.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Use zephyr-sys to get bindings

Instead of trying to manually generate the bindings, use those generated by
bindgen in the zephyr-sys crate.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Convert k_str_out to bindgen one

Remove the manual k_str_out wrapper, and use the one generated by bindgen.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
include: Add dummy field for Rust as well as CPP

For CPP builds, a few structs that end up empty in some configurations will
generate compile errors.  With the Rust tools, bindgen ends up producing
empty structsw for these as well.  Although the code compiles, it generates
warnings.  For now, add CONFIG_RUST to the ifdef checks so that these
structs don't end up empty with Rust either.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
cmake: rust: Compile the bindgen wrapper

When compiling Rust programs, the bindgen utility generates a wrapper file
to expand inline functions, as these otherwise cannot be accessed from
rust.  Pass a consistent name to the cargo build, and add it to our source
build so that this wrapper get compiled.

Signed-off-by: David Brown <david.brown@linaro.org>

teburd commented:
It's nice you got this working, it's hard to see how manually wrapping every header we have with custom bindings would have worked out. I'm interested to see how this looks in practice.

andyross commented on include/zephyr/spinlock.h:
```diff
@@ -78,7 +78,7 @@ struct k_spinlock {
 #endif /* CONFIG_SPIN_VALIDATE */
 
-#if defined(CONFIG_CPP) && !defined(CONFIG_SMP) && \
+#if (defined(CONFIG_CPP) || defined(CONFIG_RUST)) && !defined(CONFIG_SMP) && \
```
FWIW we really need to come up with a clean solution here to the "zero length spinlock" problem.  The waste of the byte (plus alignment, so generally a word) is bad enough, but the fact that the ABI changed depending on whether you want C++ (or, now, Rust) support is a pretty terrible wart.  Spinlock records in outer struts really do need to vanish[1] and not be a degenerate 1-byte struct.

That said, it's not clear to me why Rust would need this?  Does rust disallow empty struts too?

[1] On single-cpu systems without SPIN_VALIDATE, that is; i.e. the circumstances where a spinlock can be implemented with a bare arch_irq_lock/unlock().

teburd replied:
Rust is just fine with empty structs, in C its an extension and likely makes things unhappy when generating bindings

d3zd3z replied:
Rust is "fine" with it, but every struct containing one of them will generate a warning in every API call using it about it not being FFI-Safe. I'm not sure what the right solution is here.

cfriedt replied:
> FWIW we really need to come up with a clean solution here to the "zero length spinlock" problem. 

@andyross Is it possible to use a singleton for that purpose?

I actually had a bug open for that for a long time and IIRC just closed it thinking it was long forgotten / -ENOTIME.

cfriedt replied:
Does a zero-length array at the end of a struct avoid having to pad the struct with 1 byte?

d3zd3z replied:
I see a couple possible solutions here:
1. Leave the dummy byte in the spinlock.
2. Leave the empty spinlock, and suppress the warning in the generated rust code, but at a whole level. This will remove the warning, but possibly miss any other warnings about FFI issues.
3. Enhance bindgen to be able to insert the warning suppression just on specific types.

Thoughts?

d3zd3z replied:
For now, I've gone with solution 2. I've reverted the change to the C code (and will remove the patches entirely before final review), and suppressed the warning. One suggestion I did see was to add code to edit the generated bindings before using them. This is definitely possible, but a bit more awkward and fragile, and seems riskier to me than just suppressing the warning for all of the binding code. The bindings are generated from C structs so should always be valid as arguments to C functions.

andyross replied:
That sounds best to me from a general compatibility standpoint, though the C++ wart still exists and needs some genius to come along and figure out how to make this work compatibly with a clean API.  But that bikeshed isn't relevant to this PR.  +1 from me, though obviously I lack the Rust-fu to do much review on the broader PR.

andyross commented on include/zephyr/kernel/thread.h:
```diff
@@ -237,7 +237,7 @@ typedef struct k_thread_runtime_stats {
 #endif /* CONFIG_SCHED_THREAD_USAGE_ALL */
 
-#if defined(__cplusplus) && !defined(CONFIG_SCHED_THREAD_USAGE) &&                                 \
+#if (defined(__cplusplus) || CONFIG_RUST) && !defined(CONFIG_SCHED_THREAD_USAGE) &&	\
```
Same comment from elsewhere: does Rust not allow this either?  Should probably update the comment if so.

d3zd3z replied:
I'm not sure if this needs to be fixed in bindgen, or exactly what the best place to fix it is. Basically, Rust pretty readily supports zero-size items (they are used quite frequently), but considers them not FFI-safe since they aren't technically allowed by C. Maybe a longer term solution would be to inform rust that the C compiler involved supports zero length structs and this shouldn't generate this warning. It might also be possible to turn the warning off, but I would like to generate a warning with structs that are truly non FFI-safe, such as fat pointers and other features Rust has.

cfriedt replied:
C is fine, it's a C++ issue. But Rust probably tries to be safe to use with C++ 🫠

d3zd3z replied:
Actually, C is not fine with it. It is an extension that is supported, as far as I know, by both gcc and clang, but is certainly not universal. For example, apparently the MSVC compiler does not support them. That isn't likely something we need in Zephyr, but the Rust compiler intends to support interoperation with this tool

I can look at what might be possible as far as suppressing the specific warning in the Rust bindgen generated code.

d3zd3z replied:
See above, I've reverted this change, and will deal with the warning on the Rust code.

mjaun commented on lib/rust/zephyr-sys/build.rs:
```diff
@@ -0,0 +1,71 @@
+    let bindings = bindings
+        .wrap_static_fns(true)
+        .wrap_static_fns_path(wrapper_path)
+        // <inttypes.h> seems to come from somewhere mysterious in Zephyr.  For us, just pull in the
+        // one from the minimal libc.
```
The default configuration in Zephyr has `CONFIG_ENFORCE_ZEPHYR_STDINT` active which adds `-imacros${ZEPHYR_BASE}/include/zephyr/toolchain/zephyr_stdint.h`. Do we need to consider this as well?

d3zd3z replied:
I can look at pulling in this file instead of one from libc.

mjaun commented on cmake/modules/rust.cmake:
```diff
@@ -0,0 +1,185 @@
+  # themselves.
+  set(LIB_RUST_CRATES zephyr zephyr-build)
+
+  get_include_dirs(zephyr_interface include_dirs)
+
+  get_target_property(include_dirs, zephyr_interface INTERFACE_INCLUDE_DIRECTORIES)
+  get_property(include_defines TARGET zephyr_interface PROPERTY INTERFACE_COMPILE_DEFINITIONS)
```
Any reason not to use the Zephyr functions here?

```
  zephyr_get_system_include_directories_for_lang(C system_includes)
  zephyr_get_include_directories_for_lang(C includes)
  zephyr_get_compile_definitions_for_lang(C definitions)
```

d3zd3z replied:
I looked at them, but they were formatted for C. I might still be able to parse them, so I will look at it.

mjaun commented on cmake/modules/rust.cmake:
```diff
@@ -0,0 +1,185 @@
+  # Presumably, Rust applications will have no C source files, but cmake will require them.
+  # Add an empty file so that this will build.  The main will come from the rust library.
+  target_sources(app PRIVATE ${ZEPHYR_BASE}/lib/rust/main.c ${WRAPPER_FILE})
+  set_source_files_properties(
+    ${WRAPPER_FILE}
+    COMPILE_FLAGS "-I${ZEPHYR_BASE}/lib/rust/zephyr-sys"
+  )
```
If the input header file is passed as an absolute path to bindgen, the generated wrapper source file will also include it using an absolute path. Then this is not necessary.

d3zd3z commented on lib/rust/zephyr-sys/build.rs:
```diff
@@ -0,0 +1,71 @@
+        // <inttypes.h> seems to come from somewhere mysterious in Zephyr.  For us, just pull in the
+        // one from the minimal libc.
+        .clang_arg("-DRUST_BINDGEN")
+        .clang_arg("-I../../../lib/libc/minimal/include")
```
It depends a little bit on what types are being used. I would suspect that for all of the possible libc implementations basic types will be the same. In fact, there are a bunch of size checks in kernel.h that will fail if this is not the case.

d3zd3z committed:
rust: zephyr-sys: Use absolute path for wrapper

Change the header passed into bindgen to be an absolute path.  This will
cause the generated wrapper to refer to this file also using an absolute
path.  As such, remove the explicit include path added as part of the
build.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
doc: rust: Add docs on bindings

Add documentation on the bindings between Rust and C, and the bindgen
tool used to generate them.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Export zephyr-sys as zephyr::raw

Re-export all of the bindgen generated bindings in the zephyr-sys crate
into `zephyr::raw`.  This keeps things easier, as users of `zephyr` only
need to worry about the single `zephyr` crate.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
cmake: rust: Fix typo on variable name

Fix `WRAPPER_FiLE` to `WRAPPER_FILE`.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
Revert "include: Add dummy field for Rust as well as CPP"

This reverts commit 2046760e71d3cb1880c372b545d037222026ee9c.

Put these back so we get the zero element structures when using just
rust and not CPP.  A subsequent patch will suppress the warning.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Suppress warning about improper C types

Zephyr takes advantage of a gcc/clang extension that allows structs that
have no elements.  Rust is perfectly happy with this (it is quite common
in Rust code), but generates a warning when a struct containing no
elements is passed to C code.

For now, suppress this warning on the generated bindings.  This has the
disadvantage of suppressing it entirely, which might possibly detect
other cases of invalid structs.  However, the bindings are
auto-generated from C structs so should always be valid.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Workaround gcc/clang differences with soft fp

GCC automatically defines a `__SOFTFP__` define on targets that are
using software floating point.  The clang compiler does not do this by
default, so check this, and define it.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: zephyr-sys: Fixup clang targets for RISCV

Rustc for RISCV encodes optional features on the CPU available as part
of the target tuple.  Clang, on the other hand does not.  In order to be
able to use libclang with bindgen on RISCV, we need to simplify the
target tuples a bit.  Do this by just matching 'riscv32' or 'riscv64'
and then passing in a generic tuple.  We aren't generating any code, and
the structs should always match between the targets.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
.gitignore: Ignore Rust 'target' directories

Although the Cmake rules to build Rust applications keeps the target
directory inside of the build directory, some IDE tools may generate a
target directory while editing the code.  Ignore these so they never get
checked in.

Signed-off-by: David Brown <david.brown@linaro.org>

cfriedt commented:
I would guess the mps2/an521 errors might be due to stack size.

d3zd3z committed:
cmake: rust: Add header dependencies to cargo

With bindgen needing to read the headers, make sure CMake knows about
this.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z commented on cmake/modules/rust.cmake:
```diff
@@ -142,14 +168,19 @@ ${config_paths}
+  # Be sure we don't try building this until al of the generated headers have been generated.
   add_custom_target(librustapp ALL
     DEPENDS ${DUMMY_FILE}
+        # The variables, defined at the top level, don't seem to be accessible here.
```
Does anyone know how to do this right? I'd like to use the `${SYSCALL_LIST_H_TARGET}` instead of its expansion, but that variable doesn't seem to be defined at this point in this module.

mjaun replied:
`rust.cmake` is executed first, defining the `rust_cargo_application()` function. At this point `SYSCALL_LIST_H_TARGET` is not defined yet:
![image](https://github.com/user-attachments/assets/e753772d-08d0-4fe0-9ce0-50e0e1ee2f27)

Then comes the Zephyr `CMakeLists.txt` defining `SYSCALL_LIST_H_TARGET`:
![image](https://github.com/user-attachments/assets/5ed77c5d-4c88-4b58-9068-b2ea51702f82)

Finally the `rust_cargo_application()` function is executed, but the variables from Zephyr `CMakeLists.txt` are alreadyout of scope as can be seen with the stack trace:
![image](https://github.com/user-attachments/assets/0be24cd5-cc4d-47cd-8ad5-6a90663a0fa8)

I don't see how this could be done without modifying something elsewhere. Would it maybe be reasonable to just add `zephyr_interface` as a dependency?

d3zd3z replied:
I'll try adding `zephyr_interface` as a dependency, and see if it works. It's a little frustrating because the failure are somewhat non-deterministic, although it does seem to fail pretty consistently in CI.

d3zd3z replied:
So `zephyr_interface` doesn't really work as a dependency, as it seems to just hold the flags needed to set include paths for the zephyr interfaces. It doesn't actually have any header files in it. It seems that the generated targest are what is needed to have the dependencies work.

mjaun replied:
Have you tried with `add_dependencies(librustapp zephyr_interface)`? I think the `DEPENDS` parameter only works for files, not for targets.

kartben commented on doc/develop/languages/rust/bindings.rst:
```diff
@@ -0,0 +1,52 @@
+Bindings to Zephyr for Rust
+###########################
+
+Zephyr is written in C, and it's primary API is also made available to C.  It is written as a
```
```suggestion
Zephyr is written in C, and its primary API is also made available to C.  It is written as a
```

kartben commented on cmake/modules/rust.cmake:
```diff
@@ -142,14 +168,19 @@ ${config_paths}
     WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
   )
 
+  # Be sure we don't try building this until al of the generated headers have been generated.
```
```suggestion
  # Be sure we don't try building this until all of the generated headers have been generated.
```

kartben commented on doc/develop/languages/rust/bindings.rst:
```diff
@@ -0,0 +1,52 @@
+Using the Bindings
+******************
+
+In general, using direct bindings to C function from Rust is a bit more difficult than calling them
```
```suggestion
In general, using direct bindings to C functions from Rust is a bit more difficult than calling them
```

kartben commented on doc/develop/languages/rust/bindings.rst:
```diff
@@ -0,0 +1,52 @@
+
+In general, using direct bindings to C function from Rust is a bit more difficult than calling them
+from C.  Although all of these calls are considered "unsafe", and must be placed in an ``unsafe``
+block, the rust language has stricter constraints on what is allowed, even by unsafe code.  Although
```
```suggestion
block, the Rust language has stricter constraints on what is allowed, even by unsafe code.  Although
```

kartben commented on doc/develop/languages/rust/bindings.rst:
```diff
@@ -0,0 +1,52 @@
+
+In general, using direct bindings to C function from Rust is a bit more difficult than calling them
+from C.  Although all of these calls are considered "unsafe", and must be placed in an ``unsafe``
+block, the rust language has stricter constraints on what is allowed, even by unsafe code.  Although
+the intent of the Rust on Zephyr project is to allow full use of Zephyr, without needing to resort
```
I don't necessarily dislike the idea of calling this "Rust on Zephyr", but probably better to say something like "... supporting the Rust language in Zephyr ..." or something.

Also, you have two consecutive sentences starting with "Although" :)

kartben commented on cmake/modules/rust.cmake:
```diff
@@ -49,12 +49,28 @@ function(_rust_map_target)
 
+  get_include_dirs(zephyr_interface include_dirs)
+
+  get_target_property(include_dirs, zephyr_interface INTERFACE_INCLUDE_DIRECTORIES)
```
this can be dropped, I think? That's what get_include_dirs already does (L53)?

kartben commented on doc/develop/languages/rust/bindings.rst:
```diff
@@ -0,0 +1,52 @@
+
+Because the Zephyr headers use numerous conditional compilation macros, the bindings needed will be
+very specialized to a given board, and even a given configuration.  To do this, the file
+:file:`lib/rust/zephyr-sys/build.rs`, which ``cargo`` knows to run at build time will
```
```suggestion
:file:`lib/rust/zephyr-sys/build.rs`, which ``cargo`` knows to run at build time, will
```

kartben reviewed (changes requested):
some really minor comments -- note that my review concerns mostly the doc aspects of the PR

d3zd3z committed:
doc: develop: rust: Various cleanups to docs

Minor fixes to the documentation from review feedback.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
cmake: rust: Remove redundant call

Remove a redundant call that does exactly what the previous function
call accomplishes.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
cmake: rust: Comment typo fix

Fix a mispelled word "al" -> "all".

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z commented:
I'm going to go ahead an merge this, and we can focus on reviews of the cmake changes as a whole before merging the branch.
